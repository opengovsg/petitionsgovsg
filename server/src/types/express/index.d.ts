declare namespace Express {
  interface User {
    id: string
    type: string
    fullname: string
  }
}
