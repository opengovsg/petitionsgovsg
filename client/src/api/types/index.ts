export * from './post'
export * from './signature'
export * from './addressee'
export * from './auth'
