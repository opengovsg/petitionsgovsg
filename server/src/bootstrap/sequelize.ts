import { Sequelize } from 'sequelize'

import {
  defineSignature,
  definePost,
  defineAddressee,
  defineSubscription,
} from '@/models'
import { dbConfig } from './config/database'

export const sequelize = new Sequelize({ ...dbConfig, logging: false })

export const { Addressee } = defineAddressee(sequelize)
export const { Post } = definePost(sequelize, { Addressee })
export const Signature = defineSignature(sequelize, { Post })
export const Subscription = defineSubscription(sequelize, { Post })

export default sequelize
