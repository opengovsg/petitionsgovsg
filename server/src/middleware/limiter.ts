import rateLimit from 'express-rate-limit'

// Set limiter on authenticated routes to reduce burden on servers
export const limiter = rateLimit({
  windowMs: 1 * 60 * 1000, // 1 minute
  max: 10, // limit each IP to 60 requests per windowMs
  handler: function (req, res) {
    return res.status(429).json({
      error: 'You sent too many requests. Please wait a while then try again',
    })
  },
})
