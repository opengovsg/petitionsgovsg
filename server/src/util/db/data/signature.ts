export const mockSignature = {
  id: 4,
  hashedUserSgid:
    '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
  fullname: null,
  comment: 'I want to support this cause.',
  postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
  type: 'support',
}

export const mockSignatures = [
  {
    id: 1,
    hashedUserSgid: '1234',
    fullname: null,
    comment: null,
    postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    type: 'support',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 2,
    hashedUserSgid: '5678',
    fullname: 'John Doe',
    comment: null,
    postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    type: 'support',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 3,
    hashedUserSgid: '8765',
    fullname: null,
    comment: 'I support this cause!',
    postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    type: 'support',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
