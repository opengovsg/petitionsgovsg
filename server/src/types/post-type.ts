import { Post } from '~shared/types/base'

export type PostEditType = {
  id: number
  userid: number
  summary: string
  title: string
}

export type UpdatePostRequestDto = Pick<Post, 'title' | 'summary'>
