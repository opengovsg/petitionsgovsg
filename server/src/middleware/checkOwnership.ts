import { NextFunction, Request, Response } from 'express'
import { StatusCodes } from 'http-status-codes'
import { Signature, Post } from '~shared/types/base'
import { Message } from '../types/message-type'
import { ControllerHandler } from '../types/response-handler'
import { ModelDef } from '../types/sequelize'

type SignatureWithRelations = Signature & {
  userId: number
  post: Post
}
export type OwnershipCheck = ControllerHandler<
  { id: string },
  Message,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any
>
export const checkOwnershipUsing = ({
  Signature,
  Post,
}: {
  Signature: ModelDef<Signature>
  Post: ModelDef<Post>
}): OwnershipCheck => {
  const checkOwnership: OwnershipCheck = async (
    req: Request,
    res: Response,
    next: NextFunction,
  ): Promise<Response | void> => {
    //TODO

    // const results = (await Signature.findOne({
    //   where: { id: req.params.id },
    //   attributes: ['userId'],
    //   include: [
    //     {
    //       model: Post,
    //     },
    //   ],
    // })) as unknown as SignatureWithRelations
    // if (!results) {
    //   return res
    //     .status(StatusCodes.BAD_REQUEST)
    //     .json({ message: 'No answer found with this ID' })
    // }
    // if (!req.user) {
    //   return res
    //     .status(StatusCodes.UNAUTHORIZED)
    //     .json({ message: 'User not signed in' })
    // }

    // const user = await User.findByPk(req.user.id)

    // if (!user) {
    //   return res.status(StatusCodes.FORBIDDEN).json({
    //     message: `User ${req.user.id} does not exist`,
    //   })
    // }

    next()
  }
  return checkOwnership
}
