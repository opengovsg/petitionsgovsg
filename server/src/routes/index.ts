import express from 'express'
import { OwnershipCheck } from '../middleware/checkOwnership'
import { SignatureController } from '../modules/signatures/signature.controller'
import { routeSignatures } from '../modules/signatures/signature.routes'
import { AuthController } from '../modules/auth/auth.controller'
import { AuthMiddleware } from '../modules/auth/auth.middleware'
import { routeAuth } from '../modules/auth/auth.routes'
import { PostController } from '../modules/post/post.controller'
import { routePosts } from '../modules/post/post.routes'
import { routeAddressees } from '../modules/addressee/addressee.routes'
import { AddresseeController } from '../modules/addressee/addressee.controller'

type ApiRouterOptions = {
  signature: {
    controller: SignatureController
    authMiddleware: AuthMiddleware
    checkOwnership: OwnershipCheck
  }
  auth: {
    controller: AuthController
    authMiddleware: AuthMiddleware
  }
  post: {
    controller: PostController
    authMiddleware: AuthMiddleware
  }
  addressee: {
    controller: AddresseeController
  }
}

export const api = (options: ApiRouterOptions): express.Router => {
  const router = express.Router()

  router.use('/auth', routeAuth(options.auth))
  router.use('/posts', routePosts(options.post))
  router.use('/posts/signatures', routeSignatures(options.signature))
  router.use('/addressees', routeAddressees(options.addressee))

  return router
}
