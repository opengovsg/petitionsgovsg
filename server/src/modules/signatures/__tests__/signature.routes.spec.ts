import express from 'express'
import { StatusCodes } from 'http-status-codes'
import { Model, Sequelize } from 'sequelize'
import { checkOwnershipUsing } from '../../../middleware/checkOwnership'
import supertest from 'supertest'
import { Post, PostStatus, User, Signature } from '~shared/types/base'
import { ControllerHandler } from '../../../types/response-handler'
import { Creation, ModelDef } from '../../../types/sequelize'
import {
  createTestDatabase,
  getModelDef,
  ModelName,
} from '../../../util/jest-db'
import { SignatureController } from '../signature.controller'
import { routeSignatures } from '../signature.routes'
import { SignatureService } from '../signature.service'
import { okAsync } from 'neverthrow'
import { UserAuthType } from '~shared/types/api'

describe('/signatures', () => {
  let user: User
  let post: Post
  let signature: Model<Signature, Creation<Signature>> & Signature

  let db: Sequelize
  let Post: ModelDef<Post>
  let Signature: ModelDef<Signature>

  const authService = {
    hasPermissionToEditPost: jest.fn(),
  }

  // Set up auth middleware to inject user
  let authUser: Express.User | undefined = undefined
  const authenticate: ControllerHandler = (req, _res, next) => {
    req.user = authUser
    next()
  }

  const app = express()
  const request = supertest(app)

  const searchSyncService = {
    createPost: jest.fn(),
    updatePost: jest.fn(),
  }

  beforeAll(async () => {
    db = await createTestDatabase()
    Post = getModelDef<Post>(db, ModelName.Post)
    Signature = getModelDef<Signature>(db, ModelName.Signature)
    const User = getModelDef<User>(db, ModelName.User)

    const signatureService = new SignatureService({
      Post,
      Signature,
      sequelize: db,
    })
    const controller = new SignatureController({
      signatureService,
      authService,
    })

    const checkOwnership = checkOwnershipUsing({ Post, User, Signature })

    app.use(express.json())
    app.use(
      routeSignatures({
        controller,
        authMiddleware: { authenticate },
        checkOwnership,
      }),
    )

    user = await User.create({
      sgid: 'u=34',
      fullname: 'Tan Ah Wee',
      displayname: 'Mr. Tan',
      email: 'tanahwee@test.gov.sg',
      active: true,
    })
    await User.update({ id: 1 }, { where: { id: null } })
  })

  beforeEach(async () => {
    authUser = { id: user.id, type: UserAuthType.Public }
    await Post.destroy({ truncate: true })
    await Signature.destroy({ truncate: true })
    post = await Post.create({
      title: 'Post',
      summary: 'summary',
      userId: user.id,
      reason: null,
      request: null,
      references: null,
      status: PostStatus.Closed,
      fullname: 'Mr. Tan',
    })
    signature = await Signature.create({
      comment: 'Comment Body',
      userId: user.id,
      postId: post.id,
      fullname: 'Mr. Tan',
    })
    jest.resetAllMocks()
    authService.hasPermissionToEditPost.mockResolvedValue(true)
  })

  describe('GET /:id', () => {
    it('returns OK on valid query', async () => {
      const signatures = [
        {
          ...signature.toJSON(),
          createdAt: (signature.createdAt as Date).toISOString(),
          updatedAt: (signature.updatedAt as Date).toISOString(),
        },
      ]
      const response = await request.get(`/${post.id}`)

      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(signatures)
    })
  })

  describe('POST /:id', () => {
    it('returns UNAUTHORIZED on no user', async () => {
      authUser = undefined

      const response = await request
        .post(`/${post.id}`)
        .send({ text: signature.comment })

      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toStrictEqual({ message: 'User not signed in' })
    })

    it('returns OK on valid submission', async () => {
      const { id, comment: text } = signature
      await Signature.destroy({ where: { id } })

      searchSyncService.updatePost.mockResolvedValue(okAsync({}))

      const response = await request.post(`/${post.id}`).send({ text })

      const newSignatureId = Number(response.body)
      const newSignature = await Signature.findByPk(newSignatureId)

      expect(response.status).toEqual(StatusCodes.OK)
      expect(newSignature?.comment).toStrictEqual(text)
      expect(newSignature?.id).not.toBe(id)
    })
  })

  describe('DELETE /:id', () => {
    it('returns OK on valid query', async () => {
      const response = await request.delete(`/${signature.id}`)
      const deletedSignature = await Signature.findByPk(signature.id)
      expect(response.status).toEqual(StatusCodes.OK)
      expect(deletedSignature).toBeNull()
    })
  })
})
