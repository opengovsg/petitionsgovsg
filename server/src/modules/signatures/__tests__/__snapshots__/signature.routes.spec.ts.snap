// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/signatures Get /:id retrieves all signatures to a post with /:id 1`] = `
Array [
  Object {
    "comment": null,
    "createdAt": "2022-01-01T00:00:00.000Z",
    "fullname": null,
    "hashedUserSgid": "1234",
    "id": 1,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": "2022-01-01T00:00:00.000Z",
  },
  Object {
    "comment": null,
    "createdAt": "2022-01-01T00:00:00.000Z",
    "fullname": "John Doe",
    "hashedUserSgid": "5678",
    "id": 2,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": "2022-01-01T00:00:00.000Z",
  },
  Object {
    "comment": "I support this cause!",
    "createdAt": "2022-01-01T00:00:00.000Z",
    "fullname": null,
    "hashedUserSgid": "8765",
    "id": 3,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": "2022-01-01T00:00:00.000Z",
  },
]
`;

exports[`/signatures Post /:id returns BAD REQUEST on no useName 1`] = `
Object {
  "message": "useName is a required boolean",
}
`;

exports[`/signatures Post /:id returns OK on valid submission 1`] = `4`;

exports[`/signatures Post /:id returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;

exports[`/signatures Post /check/:id returns OK with null body on user has not signed 1`] = `null`;

exports[`/signatures Post /check/:id returns OK with signature on user has signed 1`] = `
Object {
  "comment": null,
  "createdAt": Any<String>,
  "fullname": "John Doe",
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "id": 4,
  "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
  "type": "support",
  "updatedAt": Any<String>,
}
`;

exports[`/signatures Post /check/:id returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;
